presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java

```diff
@@ -0,0 +1,582 @@
+        int nextPosition = positions[positionIndex];
+        int nestedIndex = 0;
+        nestedOutputPositionCount = 0;
+        for (int i = 0; i < outputPositionCount; i++) {
+            if (outputPositions[i] < nextPosition) {
+                if (!compactNulls || !nulls[i]) {
+                    nestedIndex++;
+                }
+                continue;
+            }
+
+            assert outputPositions[i] == nextPosition;
+
+            if (!compactNulls || !nulls[i]) {
+                nestedOutputPositions[nestedOutputPositionCount++] = nestedOutputPositions[nestedIndex];
+                nestedIndex++;
+            }
+            if (compactNulls) {
+                nulls[positionIndex] = nulls[i];
+            }
+            outputPositions[positionIndex] = nextPosition;
+
+            positionIndex++;
+            if (positionIndex >= positionCount) {
+                break;
+            }
+            nextPosition = positions[positionIndex];
+        }
+
+        outputPositionCount = positionCount;
+    }
+
+    private BlockLease newLease(Block block, BlockLease...fieldBlockLeases)
+    {
+        valuesInUse = true;
+        return ClosingBlockLease.newLease(block, () -> {
+            for (BlockLease lease : fieldBlockLeases) {
+                lease.close();
+            }
+            valuesInUse = false;
+        });
+    }
+
+    @Override
+    public void throwAnyError(int[] positions, int positionCount)
+    {
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .addValue(streamDescriptor)
+                .toString();
+    }
+
+    @Override
+    public void close()
+    {
+        systemMemoryContext.close();
+    }
+
+    @Override
+    public void startStripe(InputStreamSources dictionaryStreamSources, List<ColumnEncoding> encoding)
+            throws IOException
+    {
+        presentStreamSource = missingStreamSource(BooleanInputStream.class);
+
+        readOffset = 0;
+        nestedReadOffset = 0;
+
+        presentStream = null;
+
+        rowGroupOpen = false;
+
+        for (SelectiveStreamReader reader : nestedReaders.values()) {
+            reader.startStripe(dictionaryStreamSources, encoding);
+        }
+    }
+
+    @Override
+    public void startRowGroup(InputStreamSources dataStreamSources)
+            throws IOException
+    {
+        presentStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, PRESENT, BooleanInputStream.class);
+
+        readOffset = 0;
+        nestedReadOffset = 0;
+
+        presentStream = null;
+
+        rowGroupOpen = false;
+
+        for (SelectiveStreamReader reader : nestedReaders.values()) {
+            reader.startRowGroup(dataStreamSources);
+        }
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + sizeOf(outputPositions) + sizeOf(nestedPositions) + sizeOf(nestedOutputPositions) + sizeOf(nulls) +
```

Comment:
How about this?

```
                SelectiveStreamReader nestedReader = SelectiveStreamReaders.createStreamReader(
                        nestedStream,
                        ImmutableMap.of(), 
                        fieldOutputType, 
                        nestedRequiredSubfields, 
                        hiveStorageTimeZone, 
                        systemMemoryContext.newAggregatedMemoryContext());
                nestedReaders.put(fieldName, nestedReader);
```

Reply:
I considered that, but the results don't appear more readable. readerBytes doesn't work, because the value is the sum of bytes used by all the nested readers, not this reader's size in bytes. What do you think?